Add system.error_log

pamarcos opened the pull request:
Add `system.error_log` based on `system.metric_log`.
Contains history of error values from table `system.errors`, periodically flushed to disk.

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Add system.error_log which contains history of error values from table system.errors, periodically flushed to disk.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

#64501

pamarcos committed:
Remove unused headers

pamarcos committed:
Add system.error_log that flushes periodically system.errors to disk

pamarcos committed:
Remove unused imports

pamarcos committed:
Add test and config changes for system.error_log

robot-ch-test-poll commented:
<!-- automatic status comment for PR #65381 from pamarcos/ClickHouse:system-error-log -->
*This is an automated comment for commit 216d26987a12d051834d92cf15bbaf11f2e40374 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65381/216d26987a12d051834d92cf15bbaf11f2e40374/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>❌ failure</td></tr>
<tbody>
</table>

pamarcos committed:
Add documentation

pamarcos committed:
Remove more unused includes

pamarcos committed:
Fix CI style check

nikitamikhaylov commented on tests/queries/0_stateless/03172_error_log_table_not_empty.sql:
```diff
@@ -0,0 +1,5 @@
+select sleep(2) format Null; --More than collect_interval_milliseconds
```
```
select throwIf(1, 'test', toInt16(159)) settings allow_custom_error_code_in_throwif=1
```

pamarcos committed:
Fix style check

pamarcos committed:
Move ValuePair to the internal implementation

Attempt #3 to fix CI style check.
The good thing is that there was no need to expose ValuePair
to the outside.

pamarcos committed:
Fix for real the CI style check. This time is personal

pamarcos committed:
Improve error_log test to ensure errors are propagated

pamarcos committed:
Minor changes found reviewing to remove the draft status

pamarcos committed:
Fix test. Classic mistake :')

pamarcos commented on src/Interpreters/ErrorLog.cpp:
```diff
@@ -0,0 +1,127 @@
+    columns[column_idx++]->insert(remote);
+}
+
+void ErrorLog::startCollectError(size_t collect_interval_milliseconds_)
+{
+    collect_interval_milliseconds = collect_interval_milliseconds_;
+    is_shutdown_error_thread = false;
+    flush_thread = std::make_unique<ThreadFromGlobalPool>([this] { threadFunction(); });
+}
+
+
+void ErrorLog::stopCollectError()
+{
+    bool old_val = false;
+    if (!is_shutdown_error_thread.compare_exchange_strong(old_val, true))
+        return;
+    if (flush_thread)
+        flush_thread->join();
+}
+
+
+void ErrorLog::shutdown()
+{
+    stopCollectError();
+    stopFlushThread();
+}
```
This code is essentially the same as the one from `MetricLog`. I thought about consolidating in a single class this part so that both `MetricLog` and `ErrorLog` inherit from it, but I'm not sure it's worth it.

If we could consolidate also the code for the `threadFunction` itself it'd be even better, but it's not that easy to do if we want to have some state such as `previous_values` across loop iterations.

nikitamikhaylov replied:
I think this is Ok.

pamarcos replied:
I ended up consolidating the code into a `PeriodicLog` because I see there are [other issues for other periodic logs](https://github.com/ClickHouse/ClickHouse/issues/40012). So, I preferred to bite the bullet now while it's fresh

nikitamikhaylov commented on src/Interpreters/ErrorLog.cpp:
```diff
@@ -0,0 +1,127 @@
+ColumnsDescription ErrorLogElement::getColumnsDescription()
+{
+    return ColumnsDescription {
+        {"hostname", std::make_shared<DataTypeLowCardinality>(std::make_shared<DataTypeString>()), "Hostname of the server executing the query."},
+        {"event_date", std::make_shared<DataTypeDate>(), "Event date."},
+        {"event_time", std::make_shared<DataTypeDateTime>(), "Event time."},
+        {"code", std::make_shared<DataTypeInt32>(), "Error code}"},
+        {"error", std::make_shared<DataTypeLowCardinality>(std::make_shared<DataTypeString>()), "Error name."},
+        {"value", std::make_shared<DataTypeUInt64>(), "Number of errors happened in time interval."},
+        {"remote", std::make_shared<DataTypeUInt8>(), "Remote exception (i.e. received during one of the distributed queries)."}
```
Let's use compression for these columns, take a look at `getColumnsDescription` in AsynchronousMetricLog.cpp.

nikitamikhaylov commented on tests/queries/0_stateless/03172_error_log_table_not_empty.sql:
```diff
@@ -0,0 +1,25 @@
+SELECT throwIf(true, 'error_log', toInt16(333)) SETTINGS allow_custom_error_code_in_throwif=1; -- { serverError 333 }
+
+-- Wait for more than collect_interval_milliseconds to ensure system.error_log is flushed
+SELECT sleep(2) FORMAT NULL;
```
`SYSTEM FLUSH LOGS` should guarantee that logs are flushed.

pamarcos replied:
That's making sure the logs are flushed for whatever it's collected so far. But we need to ensure that something is collected, right?
I realized about this when checking the [original metric_log stateless test](https://github.com/ClickHouse/ClickHouse/blob/31eeeae4fd8bf416715f3eb9e6a70bbeec4a5deb/tests/queries/0_stateless/00990_metric_log_table_not_empty.sql)

nikitamikhaylov replied:
Ok :)

pamarcos committed:
Add ZSTD codec to system.error_log columns

pamarcos committed:
Remove another unused header

pamarcos committed:
Consolidate code into PeriodicLog

pamarcos committed:
Fix coding style

pamarcos committed:
Merge branch 'master' into system-error-log

pamarcos commented on src/Interpreters/PeriodicLog.cpp:
```diff
@@ -0,0 +1,62 @@
+void PeriodicLog<LogElement>::shutdown()
+{
+    stopCollect();
+    this->stopFlushThread();
```
Even though the `this` looks a little weird, this is to prevent a compilation error:

>error: use of undeclared identifier 'stopFlushThread'

An alternative would be to use

```c
SystemLog<LogElement>::stopFlushThread()
```
but it's much more verbose

pamarcos commented:
Checking the tests that failed. I'll be checking the boxes once the analysis of the error is finished.

- [ ] [test_checking_s3_blobs_paranoid/test.py::test_when_s3_broken_pipe_at_upload_is_retried (aarch64)](https://s3.amazonaws.com/clickhouse-test-reports/65381/71f8937abdd5d2b36c78b0ff3056f2af161325c7/integration_tests__aarch64__[5_6].html)

```=================================== FAILURES ===================================
________________ test_when_s3_broken_pipe_at_upload_is_retried _________________
[gw1] linux -- Python 3.10.12 /usr/bin/python3

cluster = <helpers.cluster.ClickHouseCluster object at 0xffa1e9b935e0>
broken_s3 = <helpers.s3_mocks.broken_s3.MockControl object at 0xffa1e939c8e0>

    def test_when_s3_broken_pipe_at_upload_is_retried(cluster, broken_s3):
        node = cluster.instances["node"]
    
        broken_s3.setup_fake_multpartuploads()
        broken_s3.setup_at_part_upload(
            count=3,
            after=2,
            action="broken_pipe",
        )
    
        insert_query_id = f"TEST_WHEN_S3_BROKEN_PIPE_AT_UPLOAD"
        node.query(
            f"""
            INSERT INTO
                TABLE FUNCTION s3(
                    'http://resolver:8083/root/data/test_when_s3_broken_pipe_at_upload_is_retried',
                    'minio', 'minio123',
                    'CSV', auto, 'none'
                )
            SELECT
                *
            FROM system.numbers
            LIMIT 1000000
            SETTINGS
                s3_max_single_part_upload_size=100,
                s3_min_upload_part_size=100000,
                s3_check_objects_after_upload=0
            """,
            query_id=insert_query_id,
        )
    
        create_multipart, upload_parts, s3_errors = get_multipart_counters(
            node, insert_query_id, log_type="QueryFinish"
        )
    
        assert create_multipart == 1
        assert upload_parts == 69
>       assert s3_errors == 3
E       assert 4 == 3

test_checking_s3_blobs_paranoid/test.py:315: AssertionError
```

- [x] Some [Integration Tests (tsan)](https://s3.amazonaws.com/clickhouse-test-reports/65381/71f8937abdd5d2b36c78b0ff3056f2af161325c7/integration_tests__tsan__[4_6].html) failed at setup due to a temporary service down

```
E               ERROR: for hdfs1  Head "https://registry-1.docker.io/v2/sequenceiq/hadoop-docker/manifests/2.7.0": received unexpected HTTP status: 503 Service Unavailable
```

- [x] [Performance comparison](https://s3.amazonaws.com/clickhouse-test-reports/65381/71f8937abdd5d2b36c78b0ff3056f2af161325c7/performance_comparison__release__[4_4]/report.html#run-errors) setup failed

```
2024-06-21 02:01:27,421: WARNING: connection: Failed to connect to localhost:9001
```

- [x] [00988_parallel_parts_removal (tsan, S3 Storage)](https://s3.amazonaws.com/clickhouse-test-reports/65381/71f8937abdd5d2b36c78b0ff3056f2af161325c7/stateless_tests__tsan__s3_storage__[1_5].html) timed out

- [x] [00763_lock_buffer_long (tsan, S3 Storage) ](https://s3.amazonaws.com/clickhouse-test-reports/65381/71f8937abdd5d2b36c78b0ff3056f2af161325c7/stateless_tests__tsan__s3_storage__[2_5].html) timed out

- [ ] [01052_window_view_proc_tumble_to_now (tsan, S3 Storage)](https://s3.amazonaws.com/clickhouse-test-reports/65381/71f8937abdd5d2b36c78b0ff3056f2af161325c7/stateless_tests__tsan__s3_storage__[2_5].html) returned something different

```
2024-06-21 03:04:04 01052_window_view_proc_tumble_to_now:                                   [ FAIL ] 129.79 sec. - result differs with reference: 
2024-06-21 03:04:04 --- /usr/share/clickhouse-test/queries/0_stateless/01052_window_view_proc_tumble_to_now.reference	2024-06-21 02:42:49.969755914 +0500
2024-06-21 03:04:04 +++ /tmp/clickhouse-test/0_stateless/01052_window_view_proc_tumble_to_now.stdout	2024-06-21 03:04:04.237564301 +0500
2024-06-21 03:04:04 @@ -1,2 +0,0 @@
2024-06-21 03:04:04 -OK
2024-06-21 03:04:04 -1
2024-06-21 03:04:04 
2024-06-21 03:04:04 
2024-06-21 03:04:04 Settings used in the test: --max_insert_threads 5 --group_by_two_level_threshold 964634 --group_by_two_level_threshold_bytes 37819916 --distributed_aggregation_memory_efficient 1 --fsync_metadata 0 --output_format_parallel_formatting 0 --input_format_parallel_parsing 1 --min_chunk_bytes_for_parallel_parsing 15395160 --max_read_buffer_size 889142 --prefer_localhost_replica 1 --max_block_size 86750 --max_joined_block_size_rows 51996 --max_threads 29 --optimize_append_index 0 --optimize_if_chain_to_multiif 0 --optimize_if_transform_strings_to_enum 1 --optimize_read_in_order 1 --optimize_or_like_chain 0 --optimize_substitute_columns 1 --enable_multiple_prewhere_read_steps 0 --read_in_order_two_level_merge_threshold 33 --optimize_aggregation_in_order 0 --aggregation_in_order_max_block_bytes 4061019 --use_uncompressed_cache 1 --min_bytes_to_use_direct_io 10737418240 --min_bytes_to_use_mmap_io 1783071701 --local_filesystem_read_method pread_threadpool --remote_filesystem_read_method read --local_filesystem_read_prefetch 1 --filesystem_cache_segments_batch_size 10 --read_from_filesystem_cache_if_exists_otherwise_bypass_cache 0 --throw_on_error_from_cache_on_write_operations 0 --remote_filesystem_read_prefetch 0 --allow_prefetched_read_pool_for_remote_filesystem 0 --filesystem_prefetch_max_memory_usage 32Mi --filesystem_prefetches_limit 10 --filesystem_prefetch_min_bytes_for_single_read_task 16Mi --filesystem_prefetch_step_marks 50 --filesystem_prefetch_step_bytes 100Mi --compile_aggregate_expressions 1 --compile_sort_description 1 --merge_tree_coarse_index_granularity 18 --optimize_distinct_in_order 0 --max_bytes_before_external_sort 5289239065 --max_bytes_before_external_group_by 10737418240 --max_bytes_before_remerge_sort 2629019942 --min_compress_block_size 1680648 --max_compress_block_size 362760 --merge_tree_compact_parts_min_granules_to_multibuffer_read 42 --optimize_sorting_by_input_stream_properties 0 --http_response_buffer_size 7633808 --http_wait_end_of_query True --enable_memory_bound_merging_of_aggregation_results 0 --min_count_to_compile_expression 3 --min_count_to_compile_aggregate_expression 3 --min_count_to_compile_sort_description 3 --session_timezone America/Mazatlan --prefer_warmed_unmerged_parts_seconds 9 --use_page_cache_for_disks_without_file_cache True --page_cache_inject_eviction False --merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability 0.59 --prefer_external_sort_block_bytes 100000000 --cross_join_min_rows_to_compress 0 --cross_join_min_bytes_to_compress 100000000 --min_external_table_block_size_bytes 0 --max_parsing_threads 1
2024-06-21 03:04:04 
2024-06-21 03:04:04 MergeTree settings used in test: --ratio_of_defaults_for_sparse_serialization 0.8483545377877905 --prefer_fetch_merged_part_size_threshold 1 --vertical_merge_algorithm_min_rows_to_activate 1 --vertical_merge_algorithm_min_columns_to_activate 69 --allow_vertical_merges_from_compact_to_wide_parts 0 --min_merge_bytes_to_use_direct_io 10737418240 --index_granularity_bytes 15547748 --merge_max_block_size 13634 --index_granularity 46160 --min_bytes_for_wide_part 140185875 --marks_compress_block_size 39946 --primary_key_compress_block_size 20714 --replace_long_file_name_to_hash 0 --max_file_name_length 0 --min_bytes_for_full_part_storage 323233178 --compact_parts_max_bytes_to_buffer 71407425 --compact_parts_max_granules_to_buffer 1 --compact_parts_merge_max_bytes_to_prefetch_part 26842182 --cache_populated_by_fetch 1 --concurrent_part_removal_threshold 4 --old_parts_lifetime 49
2024-06-21 03:04:04 
2024-06-21 03:04:04 Database: test_kgu4gf9n
```
